TwilioVerifySDK/TwilioVerify/Sources/Data/Storage+AccessGroup.swift

```diff
@@ -0,0 +1,143 @@
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+
+extension Storage {
+
+  // MARK: - Constants
+
+  enum MigrationDirection {
+    case toAccessGroup, fromAccessGroup
+  }
+
+  // MARK: - Factors
+
+  /// Returns all the stored factors excluding parameters such as accountService or accessGroup.
+  func getAllFactors(
+    using factorMapper: FactorMapperProtocol
+  ) -> [Factor] {
+    let query: CFDictionary = [
+      kSecClass: kSecClassGenericPassword,
+      kSecReturnAttributes: true,
+      kSecReturnData: true,
+      kSecMatchLimit: kSecMatchLimitAll,
+      kSecAttrAccessible: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
+    ] as CFDictionary
+
+    var result: AnyObject?
+
+    guard SecItemCopyMatching(query, &result) == errSecSuccess else {
+      Logger.shared.log(withLevel: .error, message: "Failed to obtain stored Factors.")
+      return []
+    }
+
+    let resultArray = result as? [Any] ?? []
+    
+    let data = resultArray.map {
+      (($0 as? [String: Any])?[kSecValueData as String]) as? Data
+    }.compactMap { $0 }
+
+    let factors = data.compactMap { item in
+      return try? factorMapper.fromStorage(withData: item)
+    }
+
+    return factors
+  }
+
+  /// Update Stored Factors with an specified kSecAttrAccessGroup
+  func updateFactors(
+    _ factors: [Factor],
+    with accessGroup: String
+  ) {
+    factors.forEach { factor in
+      let query = [
+        kSecClass: kSecClassGenericPassword,
+        kSecAttrAccount: factor.sid,
+        kSecAttrAccessible: kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly
+      ] as CFDictionary
+
+      let attributes: CFDictionary = [
+        kSecAttrAccessGroup: accessGroup
+      ] as CFDictionary
+
+      let status = SecItemUpdate(query , attributes)
+
+      if status == errSecSuccess {
+        Logger.shared.log(withLevel: .debug, message: "AccessGroup for factor: \(factor.sid) modified successfully")
+      } else {
+        Logger.shared.log(withLevel: .debug, message: "Unable to add accessGroup to factor: \(factor.sid) due to status: \(status)")
+      }
+
+      updateKeys(for: factor, with: accessGroup)
+    }
+  }
+
+  /// Update keyPairs with the specified accessGroup for a factor
+  private func updateKeys(
+    for factor: Factor,
+    with accessGroup: String
+  ) {
+    guard
+      let factor = factor as? PushFactor,
+      let keyPairAlias = factor.keyPairAlias,
+      let template = try? ECP256SignerTemplate(withAlias: keyPairAlias, shouldExist: true, accessGroup: accessGroup)
+    else {
+      Logger.shared.log(withLevel: .debug, message: "Unable to update keyPairs for factor: \(factor.sid)")
+      return
```

Comment:
Should we throw an error here?

Reply:
After discussing this, we prefer to not throw an error for the migration, but we should an error log and mention this edge case in the README

Reply:
Done, Thank you 👍